docs/contributing/data-handling-and-conversion.md

```diff
@@ -0,0 +1,809 @@
+
+Conceptually, the first and second items above the most problematic in the Terraform AWS Provider codebase. The first item because non-pointer types in Go cannot implement the concept of no value (`nil`). The [Zero Value Mapping section](#zero-value-mapping) will go into more details about the implications of this limitation. The second item because it can be confusing to always handle a structure ("object") type as a list.
+
+_There are efforts to replace the Terraform Plugin type system with one similar the underlying Terraform CLI type system. As these efforts materialize, this documentation will be updated._
+
+### Zero Value Mapping
+
+As mentioned in the [Type Mapping section](#type-mapping), there is a discrepency with how the Terraform Plugin SDK represents values and the reality that a Terraform State may not configure an Attribute. These values will default to the matching underlying Go type "zero value" if not set:
+
+| Terraform Plugin SDK | Go Type | Zero Value |
+|----------------------|---------|------------|
+| `TypeBool` | `bool` | `false` |
+| `TypeFloat` | `float64` | `0.0` |
+| `TypeInt` | `int` | `0` |
+| `TypeString` | `string` | `""` |
+
+For Terraform resource logic this means that these special values must always be accounted for in implementation. The semantics of the API and its meaning of the zero value will determine whether:
+
+- If it is not used/needed, then generally the zero value can safely be used to store an "unset" value and should be ignored when sending to the API.
+- If it is used/needed, whether:
+    - A value can always be set and it is safe to always send to the API. Generally, boolean values fall into this category.
+    - A different default/sentinel value must be used as the "unset" value so it can either match the default of the API or be ignored when sending to the API.
+    - A special type implementation is required within the schema to workaround the limitation.
+
+The maintainers can provide guidance on appropriate solutions for cases not mentioned in the [Recommended Implementation section](#recommended-implementations).
+
+### Root Attributes Versus Block Attributes
+
+All Attributes and Blocks at the top level of `schema.Resource` `Schema` are considered "root" attributes. These will always be handled with receiver methods on `ResourceData`, such as reading with `d.Get()`, `d.GetOk()`, etc. and writing with `d.Set()`. Any nested Attributes and Blocks inside those root Blocks will then be handled with standard Go types according to the table in the [Type Mapping section](#type-mapping).
+
+By convention in the codebase, each level of Block handling beyond root attributes should be separated into "expand" and "flatten" functions, typically named `expand{Service}{Type}` and `flatten{Service}{Type}`. The [Recommended Implementations section](#recommended-implementations) will go into those details.
```

Comment:
It would be nice to be explicit here about what an `Expand` function does vs `Flatten`

Reply:
Expanded (😂 ) on here: https://github.com/hashicorp/terraform-provider-aws/pull/16748/commits/6521344dc1169ab6f6010f0debde8598917371a8